sklearn/multioutput.py

```diff
@@ -399,7 +399,7 @@ class MultiOutputClassifier(ClassifierMixin, _MultiOutputEstimator):
     >>> X, y = make_multilabel_classification(n_classes=3, random_state=0)
     >>> clf = MultiOutputClassifier(KNeighborsClassifier()).fit(X, y)
     >>> clf.predict(X[-2:])
-    array([[1, 1, 0], [1, 1, 1]])
+    array([[1, 1, 1], [1, 1, 1]])
```

Comment:
@thomasjpfan: to follow up with , `test_kneighbors_brute_backend` bellow test the consistency of results between the two back-ends for `"brute"`.

I think this is a corner case that can be inspected.

Reply:
I haven't dived completely into this but the difference comes from `kneighbors`. Here is a simplier reproducer:

```python
import numpy as np
from sklearn.datasets import make_multilabel_classification
from sklearn.neighbors import KNeighborsClassifier

X, y = make_multilabel_classification(n_classes=3, random_state=0)
y_first = y[:, 0].reshape(-1)
knc = KNeighborsClassifier()
knc.fit(X, y_first)

knc.kneighbors(X[-2:-1])
```

This `PR`:

```bash
(array([[0.        , 6.244998  , 6.55743852, 6.63324958, 7.28010989]]),
 array([[98, 88, 29, 93, 12]]))
```

On main:

```bash
(array([[0.        , 6.244998  , 6.55743852, 6.63324958, 7.28010989]]),
 array([[98, 88, 29, 93, 63]]))
```

Note how the last index is different.

Reply:
On both branches, taking one more neighbor (6 instead of 5, the default):

```python
import numpy as np
from sklearn.datasets import make_multilabel_classification
from sklearn.neighbors import KNeighborsClassifier

X, y = make_multilabel_classification(n_classes=3, random_state=0)
y_first = y[:, 0].reshape(-1)
knc = KNeighborsClassifier()
knc.fit(X, y_first)

knc.kneighbors(X[-2:-1], n_neighbors=6)
```
gives in both cases:
```
(array([[0.        , 6.244998  , 6.55743852, 6.63324958, 7.28010989, 7.28010989]]),
 array([[98, 88, 29, 93, 63, 12]]))
 ```
 
Here, the distances associated to the samples of index 63 and 12 are equal, creating a tie.

Normally, the first candidate for the k-closest point (here 12 for k=5) must be the chosen one:

https://github.com/scikit-learn/scikit-learn/blob/305d21702d006eec3ab826e322e555ed8dfdd1ec/sklearn/neighbors/tests/test_neighbors.py#L914-L925

It seems that the new implementation follows this contract but not the one on `main`.

Must we have consistent tiebreaks for `kneighbors`? How important is this?

Personally, I would be in favor of introducing more parametrized tests for sub-classes of `PairwiseDistancesReduction`.

Reply:
> Must we have consistent tiebreaks for kneighbors? How important is this?

I think it's ok to have implementation-specific tie breaks. I think we should change the doctest to avoid having (machine precision level) tied neighbors at location k to make the doctest more robust to future in uncontrollable changes in BLAS, numpy or even hardware architectures even beyond specific implementation choices in our code-base.

I we really want to have a clean, non-arbitrary handling of ties, we could offer a randomized scan of the neighbors. This would be user-controllable by passing `random_state` constructor param (like we do for finding the best splits in decision trees for instance). By sampling different random seeds, the end-user could then build a posterior predictive distribution that reflects the uncertainty induces by tied pairwise distances between samples, with a uniform prior over tie-breaking choices. However this is likely to render the code too complex and slow. So for now I would just rewrite the test to make independent of the tie-breaking policy and keep the tie-breaking logic an implementation-specific behavior (which we could acknowledge in the docstring and/or the changelog).

Reply:
About changing the test, the fundamental problems is that `X` generated by `make_multilabel_classification` has quantized values:

```
array([[3., 6., 1., ..., 1., 5., 0.],
       [3., 5., 5., ..., 1., 1., 1.],
       [3., 3., 5., ..., 0., 2., 1.],
       ...,
       [3., 7., 3., ..., 2., 4., 2.],
       [7., 2., 1., ..., 1., 1., 1.],
       [3., 5., 3., ..., 1., 2., 4.]])
```

as a consequence computing pairwise distances in this space is full of ties... I think we can just change this doctest to use `LogisticRegression()` instead of `KNeighborsClassifier()`.